File: Object-oriented software design/Solutions/7/E1/src/com/company/EventAggregator.java
package com.company;

import java.lang.reflect.Type;
import java.util.Hashtable;
import java.util.LinkedList;
import java.util.List;

class EventAggregator {
    private static EventAggregator instance = new EventAggregator();

    private Hashtable<Type, List<Subscriber>> subscribers;

    private EventAggregator() {
        subscribers = new Hashtable<>();
    }

    void addSubscriber(Subscriber subscriber) {
        for (Type type : subscriber.getNotificationTypes()) {
            if (!subscribers.containsKey(type))
                subscribers.put(type, new LinkedList<>());

            subscribers.get(type).add(subscriber);
        }
    }

    void removeSubscriber(Subscriber subscriber) {
        for (Type type : subscriber.getNotificationTypes()) {
            if (subscribers.containsKey(type))
                subscribers.get(type).remove(subscriber);
        }
    }

    void publish(Notification notification) {
        Type type = notification.getClass();

        if (subscribers.containsKey(type))
            for (Subscriber subscriber : subscribers.get(type))
                subscriber.handle(notification);
    }

    static EventAggregator getInstance() {
        return instance;
    }
}

File: Object-oriented software design/Solutions/C/MVP.Example/MVP.Example.Refactored/EventAggregator.cs
using MVP.Example.Refactored.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MVP.Example.Refactored
{
    public interface ISubscriber<T>
    {
        void Handle(T Notification);
    }

    public class EventAggregator 
    {
        public static EventAggregator Instance = new EventAggregator();

        Dictionary<Type, List<object>> _subscribers = new Dictionary<Type, List<object>>();

        #region IEventAggregator Members

        public void AddSubscriber<T>(ISubscriber<T> Subscriber)
        {
            if (!_subscribers.ContainsKey(typeof(T)))
                _subscribers.Add(typeof(T), new List<object>());

            _subscribers[typeof(T)].Add(Subscriber);
        }

        public void RemoveSubscriber<T>(ISubscriber<T> Subscriber)
        {
            if (_subscribers.ContainsKey(typeof(T)))
                _subscribers[typeof(T)].Remove(Subscriber);
        }

      
        public void Publish<T>(T Event)
        {
            if (_subscribers.ContainsKey(typeof(T)))
                foreach (ISubscriber<T> subscriber in
                _subscribers[typeof(T)].OfType<ISubscriber<T>>())
                    subscriber.Handle(Event);
        }

        #endregion
    }

    public class PersonAddedNotification
    {
        public Person NewPerson { get; set; }
    }
}
